bctl/agent/agentObject.go

```
@@ -0,0 +1,276 @@
+
+	"bastionzero.com/bctl/v1/bctl/agent/controlchannel"
+	"bastionzero.com/bctl/v1/bctl/agent/controlchannel/agentidentity"
+	"bastionzero.com/bctl/v1/bctl/agent/controlconnection"
+	"bastionzero.com/bctl/v1/bctl/agent/keysplitting"
+	"bastionzero.com/bctl/v1/bctl/agent/registration"
+	"bastionzero.com/bctl/v1/bzerolib/connection"
+	"bastionzero.com/bctl/v1/bzerolib/connection/messenger/signalr"
+	"bastionzero.com/bctl/v1/bzerolib/connection/transporter/websocket"
+	"bastionzero.com/bctl/v1/bzerolib/logger"
+	"bastionzero.com/bctl/v1/bzerolib/messagesigner"
+	"bastionzero.com/bctl/v1/bzerolib/report"
+	"github.com/google/uuid"
+	"gopkg.in/tomb.v2"
+)
+
+const (
+	// there's nothing magical about the number 3 but we need to guarantee
+	// that the timeout is significantly larger than than the heartrate to avoid a race between receiving and reporting a pong
+	// as of this writing, this means an expected pong every minute, with a "disconnect" reported after 3 minutes
+	bastionDisconnectTimeout  = 3 * controlchannel.HeartRate
+	stoppedProcessingPongsMsg = "control channel stopped processing pongs"
+)
+
+type IRegistration interface {
+	Register(logger *logger.Logger, config registration.RegistrationConfig) error
+}
+
+type Config interface {
+	keysplitting.IKeysplittingConfig
+	registration.RegistrationConfig
+	agentidentity.IAgentIdentityTokenStore
+
+	GetTargetId() string
+	GetShutdownInfo() (string, string)
+	GetServiceUrl() string
+	GetMessageSigner() (*messagesigner.MessageSigner, error)
+
+	SetVersion(version string) error
+	SetShutdownInfo(reason string, state string) error
+}
+
+type Agent struct {
+	tmb    tomb.Tomb
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	logger *logger.Logger
+	config Config
+
+	agentType string
+
+	osSignalChan <-chan os.Signal
+	registration IRegistration
+
+	controlConn    connection.Connection
+	controlChannel *controlchannel.ControlChannel
+	version        string
+}
+
+func (a *Agent) Run(forceReRegistration bool) int {
```

I know it makes the code in agent.go:`main` a little more concise to return an int here, but I don't think it's good modularity. `Run`, for all it knows, is being called just like any other function and should return an error. It's `main`'s job to do something with that error and return the appropriate code (as is done in the daemon)